Use bulk inserts for all services during index rebuild

Alaschinger opened the pull request:
Instead of rebuilding the elastic search index entries with single index requests, now a configurable number of index request are summarized to a bulk request.
Therefor the repopulate functions of all services group multiple documents to a list now and new bundle update functions have been added to the corresponding services and to the elastic search service.

Alaschinger committed:
Use bulk inserts for all services during index rebuild

Instead of rebuilding the elastic search index entries with single index requests, now a configurable number of index request are summarized to a bulk request.
Therefor the repopulate functions of all services group multiple documents to a list now and new bundle update functions have been added to the corresponding services and to the elastic search service.

lkiesow commented on modules/elasticsearch-impl/src/main/java/org/opencastproject/elasticsearch/impl/AbstractElasticsearchIndex.java:
```diff
@@ -284,6 +286,57 @@ protected IndexResponse update(int maxRetryAttempts, int retryWaitingPeriod, Ela
+  protected BulkResponse bundleUpdate(int maxRetryAttempts, int retryWaitingPeriod,
+      List<ElasticsearchDocument> documents)
+          throws IOException, InterruptedException {
+    BulkRequest bulkRequest = new BulkRequest().setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
```
Using [`WriteRequest.RefreshPolicy.NONE` is recommended](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-refresh.html) and I don't see a reason why we would need `IMMEDIATE` here.

Alaschinger replied:
It seems that opencast doesn't display the actual version of a new created event in the adminUI if we use NONE instead of IMMEDIATE as refresh policy. The meta data fields remain empty in this case even if the ES doc has been submitted. Than an complete index rebuild would be necessary.

lkiesow commented on modules/elasticsearch-impl/src/main/java/org/opencastproject/elasticsearch/impl/AbstractElasticsearchIndex.java:
```diff
@@ -284,6 +286,57 @@ protected IndexResponse update(int maxRetryAttempts, int retryWaitingPeriod, Ela
+        } else {
+          logger.error("Could not update documents in index {} because of {}, not retrying.", getIndexName(),
+                  e.getMessage());
+          throw e;
```
Please don't separate error message and error. Throw an error with the appropriate message.

lkiesow commented on modules/series-service-impl/src/main/java/org/opencastproject/series/impl/SeriesServiceImpl.java:
```diff
@@ -765,4 +790,40 @@ private  Optional<Series> updateSeriesInIndex(String seriesId, String orgId,
+    } catch (SearchIndexException e) {
+      for (String seriesId: seriesIds) {
+        logger.error("Series {} couldn't be updated in the {} index.", seriesId, index.getIndexName(), e);
+      }
```
This will print the same error over and over again. Why not just print it once?

lkiesow commented on modules/series-service-impl/src/main/java/org/opencastproject/series/impl/SeriesServiceImpl.java:
```diff
@@ -504,53 +504,78 @@ public void repopulate() throws IndexRebuildException {
-                      updateFunctions.add(getAclUpdateFunction(seriesId, acl, organization.getId()));
-                    } catch (Exception ex) {
-                      logger.error("Unable to parse ACL of series {}.", seriesId, ex);
+      if (total > 0) {
```
This check is unnecessary, isn't it? If `total = 0` we will never run the following loop anyway.

lkiesow commented on modules/series-service-impl/src/main/java/org/opencastproject/series/impl/SeriesServiceImpl.java:
```diff
@@ -504,53 +504,78 @@ public void repopulate() throws IndexRebuildException {
+        int current = 0;
+        int n = 16;
+        while (current < total) {
+          List<SeriesEntity> seriesBundle = new ArrayList<>();
```
Naming: Please don't use names like series bundle. This is highly confusing since we work in an OSGi context and the series bundle could also mean the OSGi bundle of the series service. Maybe better to talk about a range of series, or a slice of the set of series or just series?

lkiesow commented on modules/series-service-impl/src/main/java/org/opencastproject/series/impl/SeriesServiceImpl.java:
```diff
@@ -504,53 +504,78 @@ public void repopulate() throws IndexRebuildException {
+            if (current + i >= total) {
+              break;
+            }
+            String currentOrg = databaseSeries.get(current + i).getOrganization();
+            if (currentOrg.equals(lastOrg)) {
+              seriesBundle.add(databaseSeries.get(current + i));
+            } else {
+              lastOrg = currentOrg;
+              break;
+            }
```
I think you are trying to skip series from different organizations here to only make bulk inserts for series from the same organization. But if you have series with `org1`, `org2`, `org1`, `org1` you may end up skipping the `org2` series entirely while inserting another one twice.

Plus, this also seems like a very complex way of doing this. You can probably just pass on the organization in the inserts.

lkiesow replied:
This is actually more about the complex way of passing series along to the actual update code, but this would also solve the organization issue why I put it here. Instead of passing along lists of identifiers and update functions, why not just pass on a list of data to update? That simplifies the code quite a bit and you could have something like:

```java

  var seriesBlock = new ArrayList<Series>();
  for (SeriesEntity series: seriesBundle) {
    String seriesId = series.getSeriesId();
    logger.trace("Adding series {} for organization {} to the {} index.", seriesId,
        series.getOrganization(), index.getIndexName());
    Organization organization = orgDirectory.getOrganization(series.getOrganization());
    User systemUser = SecurityUtil.createSystemUser(systemUserName, organization);
    SecurityUtil.runAs(securityService, organization, systemUser,
          () -> {
              var seriesData = Optional.of(new Series(seriesId, organization.getId()));
              try {
                DublinCoreCatalog catalog = DublinCoreXmlFormat.read(series.getDublinCoreXML());
                seriesData = getMetadataUpdateFunction(seriesId, catalog, organization.getId())
                    .apply(seriesData);
              } catch (IOException | ParserConfigurationException | SAXException e) {
                logger.error("Could not read dublincore XML of series {}.", seriesId, e);
                return;
              }

              String aclStr = series.getAccessControl();
              if (StringUtils.isNotBlank(aclStr)) {
                try {
                  AccessControlList acl = AccessControlParser.parseAcl(aclStr);
                  seriesData = getAclUpdateFunction(seriesId, acl, organization.getId())
                      .apply(seriesData);
                } catch (Exception ex) {
                  logger.error("Unable to parse ACL of series {}.", seriesId, ex);
                }
              }

              try {
                Map<String, String> properties = persistence.getSeriesProperties(seriesId);
                seriesData = getThemePropertyUpdateFunction(seriesId,
                        Optional.ofNullable(properties.get(THEME_PROPERTY_NAME)), organization.getId())
                    .apply(seriesData);
              } catch (NotFoundException | SeriesServiceDatabaseException e) {
                logger.error("Error reading properties of series {}", seriesId, e);
              }
              seriesBlock.add(seriesData.get());
          });
  }

  // do the actual index update
  updateSeriesBundleInIndex(seriesBlock);
```

You could even just run this loop over all data and then call update index if `seriesBlock.size() >= n`.

Alaschinger replied:
I just modified the original process, so that we can use multiple elements for bulk ingests. But dealing with the updated elements instead of the update functions and all the other stuff will also remove at least two additional abstraction layers. I really would prefer that and I'll try to integrate this approach in the other services, too.

lkiesow commented on modules/elasticsearch-index/src/main/java/org/opencastproject/elasticsearch/index/ElasticsearchIndex.java:
```diff
@@ -475,6 +555,85 @@ private void update(Series series) throws SearchIndexException {
+      List<Series> updatedSeriesBundle = new ArrayList<>();
+      boolean allPresent = true;
+      for (int i = 0; i < ids.size(); i++) {
+        Optional<Series> seriesOpt = getSeries(ids.get(i), orgId, user,
+                maxRetryAttemptsUpdate, retryWaitingPeriodUpdate);
```
To me, this seems really weird. We set out to rebuild the index from data we have in our database, which I think we assume to be our source of truth (compared to the index which acts as a cache). But instead of taking the data we got from the database, we start building a hyper complex way of merging database data with potentially incorrect data from the index itself?

Alaschinger replied:
I agree with you and even if that was the original way how the index has been rebuilt I'll remove that whole step, working with updated elements for the rebuild instead of old elements and update functions.

lkiesow commented on modules/series-service-impl/src/main/java/org/opencastproject/series/impl/SeriesServiceImpl.java:
```diff
@@ -504,53 +504,78 @@ public void repopulate() throws IndexRebuildException {
+                      DublinCoreCatalog catalog;
+                      try {
+                        catalog = DublinCoreXmlFormat.read(series.getDublinCoreXML());
+                        updateFunctions.add(getMetadataUpdateFunction(seriesId, catalog, organization.getId()));
```
Not sure what to do about that, and it was broken before, but this code will always set the creator to the system user, overwriting whoever actually created the series in the index.

lkiesow commented on modules/series-service-impl/src/main/java/org/opencastproject/series/impl/SeriesServiceImpl.java:
```diff
@@ -504,53 +504,78 @@ public void repopulate() throws IndexRebuildException {
+                      } catch (NotFoundException | SeriesServiceDatabaseException e) {
+                        logger.error("Error reading properties of series {}", seriesId, e);
+                      }
+                      updateFunctionBundle.add(updateFunctions.toArray(new Function[0]));
```
<del>You pass on these three update functions as three entries. Later on, that means, you get the series from the index multiple times to then run an update instead of having one update per series.</del>

Never mind. I misread that as `addAll`, you do exactly that in the next function. But why not do it in the first place? Why the array structure in between?

lkiesow commented on modules/series-service-impl/src/main/java/org/opencastproject/series/impl/SeriesServiceImpl.java:
```diff
@@ -504,53 +504,78 @@ public void repopulate() throws IndexRebuildException {
+                            updateFunctionBundle);
+
+                  });
+          for (int i = 1; i <= seriesBundle.size(); i++) {
+            int logEntry = i + (current - seriesBundle.size());
+            logIndexRebuildProgress(logger, index.getIndexName(), total, logEntry);
+          }
```
I wouldn't generate a bunch of logs simultaneously, but just log the current state instead.

lkiesow commented on modules/series-service-impl/src/main/java/org/opencastproject/series/impl/SeriesServiceImpl.java:
```diff
@@ -765,4 +790,40 @@ private  Optional<Series> updateSeriesInIndex(String seriesId, String orgId,
+
+    try {
+      List<Optional<Series>> seriesOptBundle = index.addOrUpdateSeriesBundle(seriesIds, updateFunctions, orgId, user);
+      for (String seriesId: seriesIds) {
+        logger.debug("Series {} updated in the {} index", seriesId, index.getIndexName());
+      }
```
```suggestion
      logger.debug("Series {} updated in the {} index", seriesIds, index.getIndexName());
```

lkiesow commented on modules/elasticsearch-index/src/main/java/org/opencastproject/elasticsearch/index/ElasticsearchIndex.java:
```diff
@@ -475,6 +555,85 @@ private void update(Series series) throws SearchIndexException {
   }
 
+  /**
+   * Adds or updates the series in the search index. Uses a locking mechanism to avoid issues like Lost Update.
```
> Uses a locking mechanism to avoid issues like Lost Update.

Can you elaborate on that? What do we need this locking for?

lkiesow commented on modules/elasticsearch-index/src/main/java/org/opencastproject/elasticsearch/index/ElasticsearchIndex.java:
```diff
@@ -475,6 +555,85 @@ private void update(Series series) throws SearchIndexException {
+        Optional<Series> seriesOpt = getSeries(ids.get(i), orgId, user,
+                maxRetryAttemptsUpdate, retryWaitingPeriodUpdate);
+        updatedSeriesOptBundle.add(updateFunctionBundle.get(i).apply(seriesOpt));
+        if (!updatedSeriesOptBundle.get(updatedSeriesOptBundle.size() - 1).isPresent()) {
```
Isn't this basically guaranteed to be `false` all the time, since we just run a function on this which will update or create that series? And assuming this could be `true`, that would mean, we would just skip updating the whole block of series?

Alaschinger replied:
(same as above)

lkiesow commented on modules/elasticsearch-index/src/main/java/org/opencastproject/elasticsearch/index/ElasticsearchIndex.java:
```diff
@@ -475,6 +555,85 @@ private void update(Series series) throws SearchIndexException {
+    try {
+      bundleUpdate(maxRetryAttemptsUpdate, retryWaitingPeriodUpdate, docs);
+    } catch (Throwable t) {
+      for (Series series: seriesList) {
+        throw new SearchIndexException("Cannot write series " + series + " to index", t);
+      }
```
This will throw an exception for exactly the first entry in this list. Then, throwing that exception will cause the program to immediately jump out of that loop.

lkiesow reviewed (changes requested):
I just did a review of on the series part of this pull request, but since I added quite a number of comments already, it might make sense to sort that out first and take a look at the other parts after that.

Some of my complaints are based on existing code that got reused or reworked, and we could certainly argue about if we actually want to change that or not. But I would like to (at least partly), because I find the current code extremely hard to read and understand. Mostly because of the data structures. Sometimes you have multiple lists, all containing parts of the data you need, sometimes you have the same data in multiple places at the same time, sometimes, you have lists in lists containing data and everything is named very similar (which kind of makes sense, since the variables do contain similar data).

Alaschinger committed:
simplifying & optimizing code

Removed two abstraction layers, dealing with updated elements in all services now, instead of update functions and old elements. Also done some other small fixes.

gregorydlogan commented:
IMO this looks good to go @lkiesow.  Your remaining complaints are pre-existing (and probably out of scope here anyway), or have been rebased away.  Planning on merging this prior to branch cut unless you object.
